java-plugin/src/main/java/fr/greencodeinitiative/java/checks/OptimizeImageVectorsSize.java

```diff
@@ -0,0 +1,161 @@
+import java.util.List;
+import java.util.stream.Stream;
+
+@Rule(
+        key = "EC53",
+        name = "Developpement",
+        description = OptimizeImageVectorsSize.MESSAGERULE,
+        priority = Priority.MINOR,
+        tags = {"bug"})
+public class OptimizeImageVectorsSize extends IssuableSubscriptionVisitor {
+    protected final String SVG_BEGIN = "<svg";
+    protected final String[] SVG_END = new String[]{"</svg", "</ svg>"};
+    protected final List<String> SUPERFLUOUS_ATTRIBUTES = Arrays.asList("xmlns:dc", "xmlns:cc", "xmlns:rdv", "xmlns:svg", "xmlns", "xmlns:sodipodi", "xmlns:inkscape", "inkscape:version", "inkscape:label", "inkscape:groupmode", "sodipodi:docname");
+    protected static final String MESSAGERULE = "Optimize your svg by using open source tools such as SVGGO, Compressor.io or SVG Cleaner";
+    // COMPLETE MESSAGE:
+    // "Optimize your svg by using open source tools such as SVGGO, Compressor.io or SVG Cleaner. The following should be avoided: The attribute value length can be reduced, e.g: id=\"filter4210\" to id=\"a\" (the same goes for xlink:href attribute). The attributes' value should be approximated: stdDeviation=\"0.57264819\" to stdDeviation=\"0.573\". Duplicated tags. Superfluous tags, e.g: <sodipodi:namedview>, <metadata>. Redundant tags. Superfluous attributes (xmlns:dc, xmlns:cc, xmlns:rdv, xmlns: svg, xmlns, xmlns:sodipodi, xmlns:inkscape, most of id attributes, version, inkscape:version, inkscape:label, inkscape:groupmode, sodipodi:docname)";
+
+    @Override
+    public List<Tree.Kind> nodesToVisit() {
+        return List.of(Tree.Kind.STRING_LITERAL, Tree.Kind.TEXT_BLOCK);
+    }
+
+    @Override
+    public void visitNode(Tree tree) {
+        String value = ((LiteralTree) tree).value();
+
+        // convert to lower case (not case sensitve anymore)
+        value = value.toLowerCase();
+
+        // trim a beginning and ending double quotes (") and single quotes (')
+        if (tree.is(Tree.Kind.STRING_LITERAL)) {
+            value = value.substring(1, value.length() - 1);
+        } else {
+            // a text block begins and ends with 3 double quotes
+            value = value.substring(3, value.length() - 3);
+        }
+
+        // stop escaping double quotes (") and single quotes (')
+        value = value.replaceAll("\\\\\"", "\"");
+        value = value.replaceAll("\\\\\'", "\'");
+
+        // search for svg beginning tag
+        int beginIndex = value.indexOf(SVG_BEGIN);
+        if (beginIndex < 0) {
+            // the string doesn't contain any svg
+            return;
+        }
+
+        // search for svg ending tag
+        int endIndex = -1;
+        int j = 0;
+        while (j < SVG_END.length && endIndex == -1) {
+            endIndex = value.indexOf(SVG_END[j]);
+            ++j;
+        }
+        if (endIndex == -1) {
+            return; // svg is invalid or is broken into multiple strings
+        }
+
+        // Parse svg as xml and explore its tree
```

Comment:
why do you use XML parsing and not simplier way, for example string pattern search on all java code ?

Reply:
I looked into regex matching at first, but I realized it may not suit some rules I defined.
The name of some superfluous attributes can be used as content/text instead of tags attributes, e.g: xmlns.

Reply:
Maybe we should consider using XPath? SVGs files being XML, this would allow us to perform the analysis independently of what language/framework is used. This is mentionned in the [SonarQube documentation](https://docs.sonarqube.org/latest/extension-guide/adding-coding-rules/)